### PR TITLE
Added option to Triplestore.value() to return a generator over all matching values

### PR DESCRIPTION
**Pull request by jesper-friis**
# Description
Added option to Triplestore.value() to return a generator over all matching values. Also added more tests.

**Question**: Is there a better way to indicate that we should return all matches? Using `any=None` to indicate that is not very intuitive. But using a special value of the `any` argument is logical, since returning any match (`any=True`), returning all matches (`any=None`) and require only one match (`any=False`) are mutually exclusive.

## Type of change
- [ ] Bug fix and code cleanup
- [x] New feature
- [ ] Documentation update
- [x] Testing

## Checklist for the reviewer
This checklist should be used as a help for the reviewer.

- [ ] Is the change limited to one issue?
- [ ] Does this PR close the issue?
- [ ] Is the code easy to read and understand?
- [ ] Do all new feature have an accompanying new test?
- [ ] Has the documentation been updated as necessary?
- [ ] Is the code properly tested?

### CODE REVIEW

File: tripper/triplestore.py [hunk 1]
```diff
@@ -600,33 +602,30 @@ def value(  # pylint: disable=redefined-builtin
...
+        try:
+            value = next(triples)[idx]
+        except StopIteration:
+            return default
 
         try:
             next(triples)
         except StopIteration:
-            return triple[idx]
+            return value
```
Comment (francescalb) [inline tripper/triplestore.py:625]: What is value here?

File: tripper/triplestore.py [hunk 2]
```diff
@@ -600,33 +602,30 @@ def value(  # pylint: disable=redefined-builtin
...
 
         if any:
-            return triple[idx]
+            return value
```
Comment (francescalb) [inline tripper/triplestore.py:628]: What is value here?

**Comment by francescalb** [comment]
> **Question**: Is there a better way to indicate that we should return all matches? Using `any=None` to indicate that is not very intuitive. But using a special value of the `any` argument is logical, since returning any match (`any=True`), returning all matches (`any=None`) and require only one match (`any=False`) are mutually exclusive.


How about using matches="all", "any", "unique" and have unique as default? I suppose it needs to be added in addition to any, to be deprecated.

File: tripper/triplestore.py (hunk 1 above)
Comment (jesper-friis) [reply tripper/triplestore.py:625]: It is assigned on line 618

File: tripper/triplestore.py (hunk 2 above)
Comment (jesper-friis) [reply tripper/triplestore.py:628]: > What is value here?

value is assigned on line 618

**Comment by jesper-friis** [comment]
> > **Question**: Is there a better way to indicate that we should return all matches? Using `any=None` to indicate that is not very intuitive. But using a special value of the `any` argument is logical, since returning any match (`any=True`), returning all matches (`any=None`) and require only one match (`any=False`) are mutually exclusive.
> 
> How about using matches="all", "any", "unique" and have unique as default? I suppose it needs to be added in addition to any, to be deprecated.

I like that much better, but the method is modelled after the `rdflib.Graph.value()` method that has `any` as a boolean argument. We could deprecate `any` and add `matches` as a new argument. But having an interface that follows rdflib makes it very easy to replace direct use of rdflib with tripper in existing applications. Thoughts?

File: tripper/triplestore.py [hunk 3]
```diff
@@ -580,8 +580,10 @@ def value(  # pylint: disable=redefined-builtin
             predicate: Possible criteria to match.
             object: Possible criteria to match.
             default: Value to return if no matches are found.
-            any: If true, return any matching value, otherwise raise
-                UniquenessError.
+            any: If True, return any matching value if there is more than one.
+                 If False, raise UniquenessError if there is more than one
+                 matching value.
+                 If None, return a generator over all matching values.
```
Comment (francescalb) [inline tripper/triplestore.py:586]:
```suggestion
            any: Used to define how many values to return. Argument:
                'False' (default): return the value or raise UniquenessError 
                if there is more than one matching value.             
                 'True': return any matching value if there is more than one.
                 'None': return a generator over all matching values.
```

**Comment by francescalb** [comment]
> > > **Question**: Is there a better way to indicate that we should return all matches? Using `any=None` to indicate that is not very intuitive. But using a special value of the `any` argument is logical, since returning any match (`any=True`), returning all matches (`any=None`) and require only one match (`any=False`) are mutually exclusive.
> > 
> > 
> > How about using matches="all", "any", "unique" and have unique as default? I suppose it needs to be added in addition to any, to be deprecated.
> 
> I like that much better, but the method is modelled after the `rdflib.Graph.value()` method that has `any` as a boolean argument. We could deprecate `any` and add `matches` as a new argument. But having an interface that follows rdflib makes it very easy to replace direct use of rdflib with tripper in existing applications. Thoughts?

I ma not sure, I guess it depends how important it is to stay close to rdflib. I tried to make the doicstring more understandable, but I am not sure if it worked.

**Comment by jesper-friis** [comment]
> > > > **Question**: Is there a better way to indicate that we should return all matches? Using `any=None` to indicate that is not very intuitive. But using a special value of the `any` argument is logical, since returning any match (`any=True`), returning all matches (`any=None`) and require only one match (`any=False`) are mutually exclusive.
> > > 
> > > 
> > > How about using matches="all", "any", "unique" and have unique as default? I suppose it needs to be added in addition to any, to be deprecated.
> > 
> > 
> > I like that much better, but the method is modelled after the `rdflib.Graph.value()` method that has `any` as a boolean argument. We could deprecate `any` and add `matches` as a new argument. But having an interface that follows rdflib makes it very easy to replace direct use of rdflib with tripper in existing applications. Thoughts?
> 
> I ma not sure, I guess it depends how important it is to stay close to rdflib. I tried to make the doicstring more understandable, but I am not sure if it worked.

Thanks for the improved docstring. I think that it is a good idea to stay close to rdflib, to not increase the barrier for people to use tripper in existing applications. It is still possible to add a `matches` argument in a future PR that, if given, overrides `any` without breaking backward compatibility.

**Review by francescalb — APPROVED** [review]

**Pull request merged**